### ISSUE TITLE
CUDA: Incorrect linkage with -fgpu-rdc on kernels created from lambdas inside anonymous namespaces

### ISSUE THREAD

**Issue by mkuron** [llvm/llvm-project#54560]
**Summary**

In the example below, the PTX assembly generated by Clang declares the kernel as `.weak .entry _ZN6thrust8cuda_cub4core13_kernel_agentINS0_14__parallel_for16ParallelForAgentINS0_11__transform17unary_transform_fINS_17counting_iteratorIiNS_11use_defaultES8_S8_EEPiNS5_14no_stencil_tagEZN12_GLOBAL__N_15Thing4calcESA_iNSC_5StuffEEUliE_NS5_21always_true_predicateEEElEESH_lEEvT0_T1_`. The same example compiled with NVCC generates `.weak .entry _ZN6thrust8cuda_cub4core13_kernel_agentINS0_14__parallel_for16ParallelForAgentINS0_11__transform17unary_transform_fINS_17counting_iteratorIiNS_11use_defaultES8_S8_EEPiNS5_14no_stencil_tagEZN60_GLOBAL__N__36_tmpxft_00006b02_00000000_6_b_cpp1_ii_968400945Thing4calcESA_iNSC_5StuffEEUliE_NS5_21always_true_predicateEEElEESH_lEEvT0_T1_`. In both cases, external weak linkage is used, which is not necessary since it is coming from inside a anonymous namespace. With NVCC, this is not a problem because the anonymous namespace is mangled to a unique name (`60_GLOBAL__N__36_tmpxft_00006b02_00000000_6_b_cpp1_ii_96840094`). With Clang however, the name is not unique (mangled to [`12_GLOBAL__N_1`](https://github.com/llvm/llvm-project/blob/3251ba2d0fcf5223fce3e270b91c54f548664b4e/clang/lib/AST/ItaniumMangle.cpp#L1491)). This is a problem when passing the resulting object files to `nvlink`, which will report `nvlink fatal error: Internal error: duplicate parameter bank data not same size` or `nvlink error: Duplicate weak parameter bank for ...` depending on the CUDA version. To me, it seems like internal linkage (`.entry`) instead of weak linkage (`.weak .entry`) should be used in this case.

**Versions**

I reproduced the bug with multiple Clang versions between 12.0.0 and 14.0.0. Before abd8cd9199d1e14cae961e1067b78df7044179a3 by @yxsamliu, Clang would generate `.visible .entry` instead of `.weak .entry`, which isn't any better and actually causes the example below to fail earlier on `Multiple definition of '_ZN6thrust8cuda_cub3cub11EmptyKernelIvEEvv'()`.

**Potential workaround**

It seems like the Clang option `-funique-internal-linkage-names` should be usable as a workaround that forces the symbol names to be unique, however across all Clang versions this just gives me various internal compiler errors. But that's a different issue.

**Working example**

This example uses Thrust, which makes the symbol names very lengthy, but I am pretty sure the exact same behavior can also be observed by replacing `thrust::transform` with a hand-written kernel. The lambda capture seems to be important as just putting a kernel into an anonymous namespace is not sufficient to trigger the problem.

a.cu:
```cuda
#include <thrust/transform.h>
#include <thrust/iterator/counting_iterator.h>
#include <thrust/execution_policy.h>

namespace {
    struct Stuff {
        long c[2];
    };

    struct Thing {
        void calc(int *data, int n, Stuff s) {
            auto f = [s] __device__ (int i) -> int {
                return 2*i;
            };
            
            auto first = thrust::counting_iterator<int>(0);
            auto last = first + n;
            thrust::transform(thrust::device, first, last, data, f);
        }

    };
}

void runA(int * data, int n) {
    Thing t;
    Stuff s({0, 0});
    t.calc(data, n, s);
}
```

b.cu:
```cuda
#include <thrust/transform.h>
#include <thrust/iterator/counting_iterator.h>
#include <thrust/execution_policy.h>

namespace {
    struct Stuff {
        long c[4];
    };

    struct Thing {
        void calc(int *data, int n, Stuff s) {
            auto f = [s] __device__ (int i) -> int {
                return 2*i;
            };
            
            auto first = thrust::counting_iterator<int>(0);
            auto last = first + n;
            thrust::transform(thrust::device, first, last, data, f);
        }

    };
}

void runB(int * data, int n) {
    Thing t;
    Stuff s({0, 0, 0, 0});
    t.calc(data, n, s);
}
```

Compile and link these as follows:
```bash
clang++ -O3 -std=c++14 --cuda-gpu-arch=sm_70 -fPIC -c a.cu -fgpu-rdc -o a.o
clang++ -O3 -std=c++14 --cuda-gpu-arch=sm_70 -fPIC -c b.cu -fgpu-rdc -o b.o
nvcc -O3 -Xcompiler -fPIC -gencode=arch=compute_70,code=[sm_70,compute_70] -dlink a.o b.o -o device.o
clang++ -std=c++14 -fPIC -O3 -shared -o ab.so a.o b.o device.o
```
The second-to-last command will fail with the nvlink error given in the summary.

For comparison, compile and link with NVCC:
```bash
nvcc -O3 -std=c++14 --expt-extended-lambda -gencode=arch=compute_70,code=[sm_70,compute_70] -Xcompiler -fPIC -c a.cu -dc -o a.o
nvcc -O3 -std=c++14 --expt-extended-lambda -gencode=arch=compute_70,code=[sm_70,compute_70] -Xcompiler -fPIC -c b.cu -dc -o b.o
nvcc -O3 -Xcompiler -fPIC -gencode=arch=compute_70,code=[sm_70,compute_70] -dlink a.o b.o -o device.o
nvcc -O3 -Xcompiler -fPIC -gencode=arch=compute_70,code=[sm_70,compute_70] -shared a.o b.o device.o -o ab.so
```
This will succeed.

**LLVM IR**

Unfortunately, Thrust is currently broken on clang trunk on godbolt.org, so I cannot post a link to the LLVM IR. Running clang locally, I can see that the IR uses `define weak_odr void @_ZN6thrust8cuda_cub4core13_kernel_agentINS0_14__parallel_for16ParallelForAgentINS0_11__transform17unary_transform_fINS_17counting_iteratorIiNS_11use_defaultES8_S8_EEPiNS5_14no_stencil_tagEZN12_GLOBAL__N_15Thing4calcESA_iNSC_5StuffEEUliE_NS5_21always_true_predicateEEElEESH_lEEvT0_T1_(%...`. My guess is that it should probably use some variation of `private` or `internal` instead of `weak_odr`.

**Comment by Artem-B** [comment]
It appears that we somehow failed to make the normally invisible symbol unique. That is indeed a bug.

> Minimal working example

Reproducer based on thrust is pretty hard to work with. :-(
If you could further reduce it to pure C++ code, that would help a lot.

**Comment by mkuron** [comment]
> It appears that we somehow failed to make the normally invisible symbol unique.

Do we actually have to make it unique? It should suffice to make it invisible.

> Reproducer based on thrust is pretty hard to work with. :-(

Indeed. I'm still working on getting rid of Thrust here. Hopefully I can post a more minimal example by early next week.

**Comment by yxsamliu** [comment]
It seems in general cuda-clang works for -fgpu-rdc with identical template kernel instantiation in different TU's.

This could be a corner case where the ISA of template kernel instantiated in different TU's have some difference which triggered the nvlink error. If we can dump the .bc and nvptx file of the TU's and compare these two template kernel intantiation it may help.

**Comment by mkuron** [comment]
> This could be a corner case where the ISA of template kernel instantiated in different TU's have some difference which triggered the nvlink error.

It's happening because the struct in both cases has the same name (`Stuff`, which ends up in the mangled name of the kernel), while the size of struct is different (16 byte vs. 32 byte). The size ends up in the function's parameter declaration, and that's probably what nvlink means by "parameter bank".

**Comment by Artem-B** [comment]
> Clang with -fgpu-rdc generates the non-unique .weak .entry

AFAICT clang does the right thing here, but we may still need to 'fix' it.

Let's consider https://godbolt.org/z/efGh6oa47. If the code in the anonymous namespace were to live in a header file included from multiple TUs, would we expect the final GPU executable built with -fgpu-rdc to contain just one instance of `f`, or would we get a unique instance of `f` per each TU it was used in? For C++, we'd expect to have a single instance. Each object file would have it exposed as a weak symbol and we would expect users not to violate ODR, so every function with the same name would be expected to be indeed the same function, and linker would just pick one of the instances. NVCC, on the other hand, makes each kernel unique, so we'll have all instances linked in. This is a wrong thing to do, IMO, but probably works OK for kernels as their lifetime is limited by the kernel call itself, so they can't maintain state across calls.

> `nvlink fatal error: Internal error: duplicate parameter bank data not same size`

I suspect this error may be a secondary failure. E.g. the root cause is that nvlink does not ignore the second instance of the weak symbol and fails further down the road when it needs to construct some per-kernel data which it expected to be unique. What's relevant here is that nvlink apparently fails to deal with weak kernels. I do not know whether it's by design or a bug.

Regardless of that, we'll probably need to do what NVCC does and make such kernel symbols unique, too.

> I don't understand why -dc/-fgpu-rdc even affects this kernel which cannot be invoked from outside this translation unit and thus has no obvious need of being processed by nvlink.

You do want to launch that kernel from the host and the host needs a visible GPU-side symbol to refer to, even when that symbol is not needed for linking and would normally not even be visible if it were a C++ compilation. While the kernel can not be invoked from another GPU-side TU, it does effectively get invoked from the host side of the same TU and that's what breaks the C++ visibility model. On one hand, we do expect to have the same visibility within TU on both host and device sides (hence host-side kernel stub needs to see GPU-side kernel symbol), but we also need to keep GPU objects linkable and due to the way `nvlink` apparently works, we do need to make those symbols unique in order to avoid name conflicts.

**Comment by mkuron** [comment]
> If the code in the anonymous namespace were to live in a header file included from multiple TUs, would we expect the final GPU executable built with -fgpu-rdc to contain just one instance of f, or would we get a unique instance of f per each TU it was used in? For C++, we'd expect to have a single instance. [...] NVCC, on the other hand, makes each kernel unique, so we'll have all instances linked in. This is a wrong thing to do, IMO, but probably works OK for kernels as their lifetime is limited by the kernel call itself, so they can't maintain state across calls.

Ah, I forgot about that scenario. So that means that as long as CUDA has no concept of visibility, it is not possible to correctly represent this C++ as PTX. If we want to match NVCC and support my scenario, we need to make kernel names unique. We can keep the `.weak` for your scenario, but it won't do much because the kernels will have different names between files.

> I suspect this error may be a secondary failure. E.g. the root cause is that nvlink does not ignore the second instance of the weak symbol and fails further down the road when it needs to construct some per-kernel data which it expected to be unique. What's relevant here is that nvlink apparently fails to deal with weak kernels. I do not know whether it's by design or a bug.

The weak kernels with the non-unique name are actually different in my example. nvlink however only complains about that because their signatures are also different:

```asm
.weak .entry _ZN12_GLOBAL__N_11fEPiiNS_5StuffE(
        .param .u64 _ZN12_GLOBAL__N_11fEPiiNS_5StuffE_param_0,
        .param .u32 _ZN12_GLOBAL__N_11fEPiiNS_5StuffE_param_1,
        .param .align 8 .b8 _ZN12_GLOBAL__N_11fEPiiNS_5StuffE_param_2[16]
)
.weak .entry _ZN12_GLOBAL__N_11fEPiiNS_5StuffE(
        .param .u64 _ZN12_GLOBAL__N_11fEPiiNS_5StuffE_param_0,
        .param .u32 _ZN12_GLOBAL__N_11fEPiiNS_5StuffE_param_1,
        .param .align 8 .b8 _ZN12_GLOBAL__N_11fEPiiNS_5StuffE_param_2[32]
)
```
If the signature is the same (e.g. if the `Stuff` struct is the same size and only the function `f` is different between the two files), nvlink does not complain, even though the generated PTX of the two files now contains two different weak implementations of the same kernel. I don't know whether the runtime will launch the right kernel when asked to -- if it does so via the name, it probably won't.

**Comment by yxsamliu** [comment]
In C++ program, for functions in an anonymous namespace, clang emits the function with internal linkage and a non-unique name. This indicates functions with the same name in the anonymous namespace in different TU's are not treated as one definition.

For CUDA/HIP, we have to make kernel symbols visible to runtime, therefore kernels must have non-internal linkage. To avoid conflict with kernels with the same name in other TU's, we have to make the kernel name unique. We also need to let the host compilation know the unique name used in the device compilation.

This is a similar situation we faced with making static device variables accessible to host compilation. We need to make their name unique and at the same time the unique name needs to be known at host compilation. We introduced CUID and externalized static device variables to solve the issue. It seems we could do that again with kernels in an anonymous namespace.

BTW, we need to externalize device variables in an anonymous namespace too.

**Comment by yxsamliu** [comment]
I have a patch for fixing this issue: https://reviews.llvm.org/D123353

**Comment by mkuron** [comment]
Sorry for the late reply, I finally got around to testing the patch today. Unfortunately, it does not (fully) solve the issue. @yxsamliu, could you please reopen this ticket?

There are now two problems:

1. The patch has no effect on the code samples I provided above. They still yield `duplicate parameter bank data not same size` errors because the name is not made unique.
2. The patch does have the intended effect on a simpler code sample (below, without templates). However, the generated PTX is invalid: `.weak .entry _ZN12_GLOBAL__N_11_fEPiiNS_5StuffE.anon.3993e450b33cd48`, which leads to `ptxas ... fatal : Parsing error near '.anon': syntax error`. As per 5db24d7c2290f8b1eabb29ae6708d6ec34ebceaa, dots are not valid in PTX identifiers.

```cuda
namespace {
    struct Stuff {
        long c[2];
    };

    __global__ void f(int * data, int n, Stuff s) {
        if(threadIdx.x < n) {
            data[threadIdx.x] = 2*threadIdx.x;
        }
    }

    void calc(int *data, int n, Stuff s) {
        f<<<n/256+1, 256>>>(data, n, s);
    }
}

void run(int * data, int n, Stuff s) {
    calc(data, n, s);
}
```

**Comment by yxsamliu** [comment]
The second issue is easy fix.

For the first issue, it seems a template instantiation with template argument type in anonymous name space will have internal linkage in C++, even though the template itself is not in anonymous name space. This is in contrary with ordinary template instantiation, which results in linkonce_odr linkage (https://godbolt.org/z/q765ehqxv). Therefore template instantiation with argument type in anonymous namespace does not follow ODR even if the type name is the same. Therefore we are justifiable to make them unique in CUDA/HIP. I will fix it.

**Comment by yxsamliu** [comment]
Fixed in https://reviews.llvm.org/D124189

**Comment by mkuron** [comment]
@yxsamliu, thanks for taking care of this so quickly. I have confirmed that your latest patch resolves the issue with the above pieces of sample code and also with the larger codebase that initially led me to discover the issue.

I hope you can merge this bugfix to master soon. What would it take to get it backported to the 14.x branch as well?

**Comment by yxsamliu** [comment]
@tstellar what's the procedure for cherrypicking https://reviews.llvm.org/rG4ea1d435099f992cc16127619b0feb64e070630d and  https://reviews.llvm.org/rG04fb81674ed7981397ffe70fe6a07b7168f6fe2f to 14.x ? Thanks.

**Comment by tstellar** [comment]
/cherry-pick 4ea1d435099f992cc16127619b0feb64e070630d 04fb81674ed7981397ffe70fe6a07b7168f6fe2f

**Comment by tstellar** [comment]
@yxsamliu There are test failures with your branch.

**Comment by yxsamliu** [comment]
> @yxsamliu There are test failures with your branch.

You mean https://github.com/llvmbot/llvm-project/pull/187 ? It should be closed since @mkuron has another PR that is working. I don't have permission to close https://github.com/llvmbot/llvm-project/pull/187

**Comment by mkuron** [comment]
@tstellar, please use https://github.com/llvmbot/llvm-project/pull/185 instead of https://github.com/llvmbot/llvm-project/pull/187. On the former, I‘ve fixed the ABI incompatibility.

**Comment by tstellar** [comment]
Merged: 29f1039a7285a5c3a9c353d054140bf2556d4c4d
